File: windows/Tokendial.Core/Install/InstallWatcher.cs
namespace Tokendial.Core.Install;

public enum InstallOutcome
{
    /// <summary>The tool signed in; the provider can be connected.</summary>
    Success,
    /// <summary>Nothing signed in within the wait.</summary>
    TimedOut,
    /// <summary>The terminal closed before a sign-in appeared.</summary>
    TerminalClosed
}

/// <summary>
/// Follows one provider after its terminal opened: not installed → installed → signed in, on a clock,
/// and says how it ended. Thread-agnostic like the store, with no timer of its own in tests; it touches
/// neither settings nor the store, it only raises events.
/// </summary>
public sealed class InstallWatcher : IDisposable
{
    public static readonly TimeSpan DefaultTimeout = TimeSpan.FromMinutes(15);
    public static readonly TimeSpan DefaultInterval = TimeSpan.FromSeconds(3);

    private readonly object gate = new();
    private readonly Func<bool> installed;
    private readonly Func<bool> signedIn;
    private readonly Func<DateTimeOffset> now;
    private readonly TimeSpan timeout;
    private readonly bool waitsAfterTerminal;
    private readonly DateTimeOffset started;
    private Timer? timer;
    private bool done;

    /// <param name="waitsAfterTerminal">An app signs in on its own later, so closing the terminal does not end the wait once it is installed.</param>
    public InstallWatcher(string providerId, Func<bool> installed, Func<bool> signedIn, bool waitsAfterTerminal = false, Func<DateTimeOffset>? now = null, TimeSpan? timeout = null)
    {
        ProviderId = providerId;
        this.installed = installed;
        this.signedIn = signedIn;
        this.waitsAfterTerminal = waitsAfterTerminal;
        this.now = now ?? (() => DateTimeOffset.UtcNow);
        this.timeout = timeout ?? DefaultTimeout;
        started = this.now();
        State = installed() ? InstallState.Installed : InstallState.NotInstalled;
    }

    public string ProviderId { get; }
    public InstallState State { get; private set; }
    public InstallOutcome? Outcome { get; private set; }
    public bool Running => Outcome is null;

    public event Action<InstallState>? Advanced;
    public event Action<InstallOutcome>? Completed;

    public void Start(TimeSpan? interval = null)
    {
        var every = interval ?? DefaultInterval;
        timer = new Timer(_ => Tick(), null, every, every);
    }

    public void Tick()
    {
        InstallState? advanced = null;
        InstallOutcome? outcome = null;
        lock (gate)
        {
            if (done) return;
            if (signedIn())
            {
                State = InstallState.SignedIn;
                advanced = State;
                outcome = InstallOutcome.Success;
            }
            else
            {
                if (State == InstallState.NotInstalled && installed())
                {
                    State = InstallState.Installed;
                    advanced = State;
                }
                if (now() - started >= timeout) outcome = InstallOutcome.TimedOut;
            }
            if (outcome is not null) Finish(outcome.Value);
        }
        if (advanced is InstallState state) Advanced?.Invoke(state);
        if (outcome is InstallOutcome ended) Completed?.Invoke(ended);
    }

    /// <summary>The terminal window closed. A CLI signs in inside it, so closing it early ends the wait; an installed app signs in on its own later, so the wait goes on.</summary>
    public void TerminalClosed()
    {
        Tick();
        bool ended;
        lock (gate)
        {
            ended = !done && !(waitsAfterTerminal && State == InstallState.Installed);
            if (ended) Finish(InstallOutcome.TerminalClosed);
        }
        if (ended) Completed?.Invoke(InstallOutcome.TerminalClosed);
    }

    private void Finish(InstallOutcome outcome)
    {
        done = true;
        Outcome = outcome;
        timer?.Dispose();
        timer = null;
    }

    public void Dispose()
    {
        lock (gate)
        {
            done = true;
            timer?.Dispose();
            timer = null;
        }
    }
}

File: macos/TokendialCore/Sources/TokendialCore/Install/InstallWatcher.swift
import Foundation

public enum InstallOutcome: Sendable {
    case success
    case timedOut
    case terminalClosed
}

/// Watches one install through: first the tool appears on disk, then its credential does. It polls, like the
/// session monitors, because none of what it waits for announces itself.
///
/// One difference from Windows, and it is forced: there the terminal is a child process, so its exit is an
/// event. Here the terminal belongs to the user, so the script drops a marker file on its way out and this
/// watches for that instead. A sign-in that lands in the same tick still wins, so closing the window right
/// after finishing does not lose the result.
public final class InstallWatcher {
    public static let defaultTimeout: TimeInterval = 15 * 60
    public static let defaultInterval: TimeInterval = 3

    public let providerId: String
    public private(set) var state: InstallState
    public private(set) var outcome: InstallOutcome?

    public var advanced: ((InstallState) -> Void)?
    public var completed: ((InstallOutcome) -> Void)?

    private let installed: () -> Bool
    private let signedIn: () -> Bool
    private let terminalFinished: () -> Bool
    private let waitsAfterTerminal: Bool
    private let timeout: TimeInterval
    private let now: () -> Date
    private let lock = NSLock()
    private var timer: DispatchSourceTimer?
    private var started: Date

    public init(providerId: String,
                installed: @escaping () -> Bool,
                signedIn: @escaping () -> Bool,
                terminalFinished: @escaping () -> Bool = { false },
                waitsAfterTerminal: Bool = false,
                now: @escaping () -> Date = Date.init,
                timeout: TimeInterval = InstallWatcher.defaultTimeout) {
        self.providerId = providerId
        self.installed = installed
        self.signedIn = signedIn
        self.terminalFinished = terminalFinished
        self.waitsAfterTerminal = waitsAfterTerminal
        self.now = now
        self.timeout = timeout
        self.started = now()
        self.state = installed() ? .installed : .notInstalled
    }

    public var running: Bool { lock.withLock { outcome == nil } }

    public func start(interval: TimeInterval = InstallWatcher.defaultInterval) {
        lock.withLock { started = now() }
        let timer = DispatchSource.makeTimerSource(queue: .global(qos: .utility))
        timer.schedule(deadline: .now() + interval, repeating: interval)
        timer.setEventHandler { [weak self] in self?.tick() }
        lock.withLock { self.timer = timer }
        timer.resume()
    }

    /// One look at the world. Public so a test can drive it without a clock.
    public func tick() {
        var advance: InstallState?
        var finish: InstallOutcome?
        lock.lock()
        if outcome == nil {
            if signedIn() {
                if state != .signedIn { state = .signedIn; advance = state }
                finish = .success
            } else if state == .notInstalled, installed() {
                state = .installed
                advance = state
            }
            if finish == nil, terminalFinished(), !(waitsAfterTerminal && state == .installed) {
                finish = .terminalClosed
            }
            if finish == nil, now().timeIntervalSince(started) >= timeout {
                finish = .timedOut
            }
            if let finish {
                outcome = finish
                timer?.cancel()
                timer = nil
            }
        }
        lock.unlock()
        if let advance { advanced?(advance) }
        if let finish { completed?(finish) }
    }

    public func stop() {
        lock.withLock {
            timer?.cancel()
            timer = nil
        }
    }
}
